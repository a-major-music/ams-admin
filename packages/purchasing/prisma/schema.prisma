// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
  output   = "./generated/client"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Address {
  id         Int       @id @default(autoincrement())
  guid       String    @unique @default(uuid())
  address1   String
  address2   String?
  address3   String?
  city       String?
  postcode   String?
  supplier   Supplier? @relation(fields: [supplierId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  supplierId Int       @unique
}

model Contact {
  id         Int      @id @default(autoincrement())
  guid       String   @unique @default(uuid())
  name       String
  email      String?
  phone      String?
  supplier   Supplier @relation(fields: [supplierId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  supplierId Int      @unique
}

model Supplier {
  id             Int             @id @default(autoincrement())
  guid           String          @unique @default(uuid())
  name           String          @unique
  website        String?
  address        Address?
  contacts       Contact[]
  purchaseOrders PurchaseOrder[]
}

model PurchaseOrder {
  id          Int                     @id @default(autoincrement())
  guid        String                  @unique @default(uuid())
  number      String                  @unique
  state       String                  @default("draft")
  issueDate   DateTime?
  pendingXero Boolean                 @default(true)
  lineItems   PurchaseOrderLineItem[]
  subTotal    Int
  taxAmount   Int
  total       Int
  supplier    Supplier?               @relation(fields: [supplierId], references: [id])
  supplierId  Int?
}

model PurchaseOrderLineItem {
  id                  Int                            @id @default(autoincrement())
  guid                String                         @unique @default(uuid())
  variantSku          String?
  variantGuid         String?
  variantUPC          String?
  variantSupplierCode String?
  variantTaxRate      Int
  description         String
  variantMediaURI     String?
  itemCost            Int
  quantityOrdered     Int
  packSize            Int @default(1)
  receipts            PurchaseOrderLineItemReceipt[]
  purchaseOrder       PurchaseOrder                  @relation(fields: [purchaseOrderId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  purchaseOrderId     Int
}

model PurchaseOrderLineItemReceipt {
  id                      Int                    @id @default(autoincrement())
  guid                    String                 @unique @default(uuid())
  date                    DateTime               @default(now())
  received                Int
  receivedPrice           Int
  PurchaseOrderLineItem   PurchaseOrderLineItem? @relation(fields: [purchaseOrderLineItemId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  purchaseOrderLineItemId Int?
}
