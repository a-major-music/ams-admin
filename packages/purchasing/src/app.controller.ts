import { Res } from "@nestjs/common";
import { Controller, Get } from "@nestjs/common";
import { ApiTags, ApiOperation } from "@nestjs/swagger";
import { AppService } from "./app.service";

@ApiTags("Main")
@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get()
  @ApiOperation({ description: "Loads this swagger page" })
  redirect(@Res() res) {
    return res.redirect("/swagger-ui");
  }
}
