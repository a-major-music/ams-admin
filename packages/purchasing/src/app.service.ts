import { Injectable } from '@nestjs/common';

@Injectable()
export class AppService {
  getHello(): string {
    return 'Hello World! <a href=/swagger-ui>swagger-ui</a>';
  }
}
