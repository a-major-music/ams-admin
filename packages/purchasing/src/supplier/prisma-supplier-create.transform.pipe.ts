import { PipeTransform, Injectable, ArgumentMetadata } from "@nestjs/common";

import _ = require("lodash");

/* 
The types defined in @amm/types are related, and these relationships mirror the structure of the database. However,
Prisma does not use DTOs and repositories like other ORMs. Instead it uses a JSON format to define both data and
operation in the same structure. Top level items are the same, but nested objects need operations inserted in the
object tree to tell Prisma what to do. So a create({id: 1, nested: {id: 1}}) becomes 
create({id: 1, nested: create: {id: 1}}). This interceptor inserts these operations into the objects before they hit
the controller to create variants, and to connectOrCreate suppliers and tags (to avoid duplicates)
*/
@Injectable()
export class PrismaSupplierCreateTransformPipe implements PipeTransform {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  transform(value: any, metadata: ArgumentMetadata) {
    const augments: any = {};

    if (!_.isEmpty(value.contacts)) {
      augments.contacts = {
        create: value.contacts,
      };
    }

    if (!_.isEmpty(value.address)) {
      augments.address = {
        create: value.address,
      };
    }

    return { ...value, ...augments };
  }
}
