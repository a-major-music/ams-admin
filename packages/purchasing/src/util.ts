export function removeProps(obj: object, props: string[]): object {
  return JSON.parse(
    JSON.stringify(obj, (k, v) => (props.includes(k) ? undefined : v))
  );
}
