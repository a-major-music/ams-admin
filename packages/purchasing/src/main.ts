import { ConfigService } from "@nestjs/config";
import { NestFactory, Reflector } from "@nestjs/core";
import { SwaggerModule, DocumentBuilder } from "@nestjs/swagger";
import { AppModule } from "./app.module";
import * as bodyParser from "body-parser";
import { Auth0JwtGuard, DomainAuthGuard } from "../../common/src/auth";

async function bootstrap() {
  const app = await NestFactory.create(AppModule);

  const config = new DocumentBuilder()
    .setTitle("AMM - Purchasing Services")
    .setDescription("The Inventory for AAM Purchasing Services")
    .setVersion("1.0")
    .addTag("purchasing", "API Endpoint for purchasing")
    .addTag("supplier", "API Endpoint for supplier")
    .build();
  const document = SwaggerModule.createDocument(app, config);
  SwaggerModule.setup("swagger-ui", app, document);

  const configService: ConfigService = app.get(ConfigService);
  app.use(bodyParser.json({ limit: "50mb" }));
  app.use(bodyParser.urlencoded({ limit: "50mb", extended: true }));
  app.enableCors();
  const auth0Guard = app.get(Auth0JwtGuard);
  const domainGuard = app.get(DomainAuthGuard);
  
  app.useGlobalGuards(auth0Guard, domainGuard);
  await app.listen(configService.get("PORT") || 3001);
}
bootstrap();
