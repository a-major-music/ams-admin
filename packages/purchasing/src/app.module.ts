import { Module } from "@nestjs/common";
import { ConfigModule } from "@nestjs/config";
import { AppController } from "./app.controller";
import { AppService } from "./app.service";
import { PurchasingModule } from "./purchasing/purchasing.module";
import { SupplierModule } from "./supplier/supplier.module";
import { AmmDocModule } from "@amm/amm-doc";
import { EventEmitterModule } from "@nestjs/event-emitter";
import { AuthModule } from "../../common/src/auth/auth.module";

@Module({
  imports: [
    ConfigModule.forRoot(),
    AuthModule,
    PurchasingModule,
    SupplierModule,
    AmmDocModule,
    EventEmitterModule.forRoot(),
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
