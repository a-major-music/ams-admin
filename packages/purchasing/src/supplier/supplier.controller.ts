import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Post,
  Put,
  Query,
} from "@nestjs/common";
import { Prisma, Supplier } from "../../prisma/generated/client";
import { ApiOperation, ApiParam } from '@nestjs/swagger';
import { PrismaSupplierCreateTransformPipe } from "./prisma-supplier-create.transform.pipe";
import { SupplierService } from "./supplier.service";

@Controller("supplier")
export class SupplierController {
  constructor(private supplierService: SupplierService) {}

  @Get('count')
  async countSuppliers(): Promise<any> {
    return { rows: (await this.supplierService.countSuppliers())._count.id };
  }

  @Get(":guid")
  async findByGuid(@Param("guid") guid: string): Promise<Supplier> {
    return this.supplierService.findByGuid(guid);
  }

  @Get()
  @ApiOperation({
    description:
      'Returns all suppliers as array of Supplier, with deep fetch of nested objects',
  })
  @ApiParam({
    name: 'skip',
    required: false,
    description: 'number of records to skip',
    type: 'integer',
  })
  @ApiParam({
    name: 'take',
    required: false,
    description: 'number of records to return',
    type: 'integer',
  })
  @ApiParam({
    name: 'where',
    required: false,
    description: 'optional where clause',
    type: 'object',
  })
  @ApiParam({
    name: 'orderBy',
    required: false,
    description: 'optional orderBy clause',
    type: 'object',
  })
  async findAll(
    @Query('skip') skip?: number,
    @Query('take') take?: number,
    @Query('where') where?: Prisma.SupplierWhereInput,
    @Query('orderBy') orderBy? : Prisma.SupplierOrderByWithRelationInput
  ): Promise<Supplier[]> {
    const params: any = {};
    if (skip) params.skip = +skip;
    if (take) params.take = +take;
    if (where) params.where = where;
    if (orderBy) params.orderBy = orderBy;

    return this.supplierService.findMany(params);
  }

  @Post("create")
  async create(
    @Body(PrismaSupplierCreateTransformPipe) supplier: Supplier
  ): Promise<Supplier> {
    return this.supplierService.create(supplier).catch((e) => {
      console.error(e);
      throw new HttpException(e, HttpStatus.BAD_REQUEST);
    });
  }

  @Put(":guid/update")
  async update(
    @Param("guid") guid: string,
    @Body() data: Supplier
  ): Promise<any> {
    const params = { where: { guid: guid }, data: data };
    return this.supplierService.update(params);
  }

  @Delete(":guid/delete")
  async delete(@Param("guid") guid: string): Promise<Supplier> {
    return this.supplierService.delete(guid);
  }
}
