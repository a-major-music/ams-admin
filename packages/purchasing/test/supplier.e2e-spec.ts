import { INestApplication } from "@nestjs/common";
import { Test } from "@nestjs/testing";
import { AppModule } from "../src/app.module";
import { SupplierModule } from "../src/supplier/supplier.module";

import supertest = require("supertest");
import { PrismaService } from "../src/prisma.service";
import { removeProps } from "../src/util";
import { createMock } from "@golevelup/ts-jest";

import { Address } from "@amm/types";

let app: INestApplication;
let prisma: PrismaService;

beforeAll(async () => {
  const module = await Test.createTestingModule({
    imports: [AppModule, SupplierModule],
  }).compile();

  app = module.createNestApplication();
  prisma = app.get<PrismaService>(PrismaService);
  await app.init();
});

afterAll(async () => {
  await app.close();
});

describe("POST /supplier/create", () => {
  it("should create a single supplier", async () => {
    const s = {
      name: "Testing create supplier",
      website: "https://nowhere.com",
      address: createMock<Address>(),
      contacts: [
        {
          name: "Testing create supplier (contact)",
        },
      ],
    };

    const cs = await supertest
      .agent(app.getHttpServer())
      .post("/supplier/create")
      .set("Accept", "application/json")
      .send(s)
      .expect("Content-Type", /json/)
      .expect(201);

    const qs = await supertest
      .agent(app.getHttpServer())
      .get(`/supplier/${cs.body.guid}`)
      .set("Accept", "application/json")
      .expect("Content-Type", /json/)
      .expect(200);

    expect(removeProps(qs.body, ["guid"])).toMatchSnapshot();

    await prisma.address.deleteMany({});
    await prisma.contact.deleteMany({});
    await prisma.supplier.deleteMany({});
  });
});
