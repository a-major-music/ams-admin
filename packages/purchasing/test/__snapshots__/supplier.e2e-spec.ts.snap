// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /supplier/create should create a single supplier 1`] = `
Object {
  "address": null,
  "contacts": Array [
    Object {
      "email": null,
      "id": 1,
      "name": "Testing create supplier (contact)",
      "phone": null,
      "supplierId": 4,
    },
  ],
  "id": 4,
  "name": "Testing create supplier",
  "website": "https://nowhere.com",
}
`;
